### DISCUSSION TITLE
5 axis gantry machine with cloned Y axis for autosquaring.

### DISCUSSION THREAD

**Discussion by mlinka** [fra589/grbl-Mega-5X#263] [Q&A] [answered]
Hello and thank you for your great work.
I am in the process of connecting the hardware to RAMPS 1.6
The machine is 3 axis gantry with 2 motors on Y + 4th and 5th axis.
I would like to have Auto squaring homing option on Y axis. Can you point me in the right direction on how to connect the Homing switches and setting the GRBL? 
Also is there a way to use some kind of macro for ATC, or how would I implement ATC option of my spindle?
(The tools will be arranged in a row on X axis - 8 pcs)
I am attaching a picture how I planned to connect the RAMPS board to motors.
Thank you very much and sorry for my english.
Best regards.
![MEGA 5X](https://user-images.githubusercontent.com/12416819/158380736-78478814-92db-4bbf-a77f-2936128588b6.jpg)
.

**Comment by mlinka** [comment]
My axis definitions:

//----------------------------------------------------------------------
// Axis definitions :
//----------------------------------------------------------------------
// IMPORTANT: When changing the axis definitions (axis numbers N_AXIS, 
// linears axis number N_AXIS_LINEAR or axes names AXIS_*_NAME, 
// don't forget to issue the reset factory defaults Grbl command: $RST=*
// if you forget the $RST=* command after change, Grbl may have 
// unpredictable behavior!
//----------------------------------------------------------------------
#define N_AXIS 6        // Number of axes (3 to 6)
#define N_AXIS_LINEAR 4 // Number of linears axis, must be <= N_AXIS
// Axis indexing and names
#define AXIS_1 0        // Axis indexing value. Must start with 0 and be continuous.
#define AXIS_1_NAME 'X' // Axis names must be in X, Y, Z, A, B, C, U, V, W, D, E & H.
#define AXIS_2 1
#define AXIS_2_NAME 'Y'
#define AXIS_3 2
#define AXIS_3_NAME 'Z'
#if N_AXIS <3
  #error "N_AXIS must be >= 3. N_AXIS < 3 is not implemented."
#endif
#if N_AXIS > 3
  #define AXIS_4 3
  #define AXIS_4_NAME 'A' // Letter of axis number 4
#endif
#if N_AXIS > 4
  #define AXIS_5 4
  #define AXIS_5_NAME 'C' // Letter of axis number 5
#endif
#if N_AXIS > 5
  #define AXIS_6 5
  #define AXIS_6_NAME 'Y' // Letter of axis number 6
#endif
#if N_AXIS > 6
  #error "N_AXIS must be <= 6. N_AXIS > 6 is not implemented."
#endif
//----------------------------------------------------------------------

**Comment by fra589** [comment, marked as answer]
Hello @mlinka,  
  
Nice project!  
  
Warning, due to internal algorithm, linear axis numbers must be contiguous and lower than rotational axes.
So, you must use axis numbers 5 and 6 for your A and C axes.
I recommend you to use those definition:
```(...)
#define N_AXIS 6 // Number of axes (3 to 6)
#define N_AXIS_LINEAR 4 // Number of linear axes, must be <= N_AXIS
(...)
#define AXIS_1_NAME 'X'
(...)
#define AXIS_2_NAME 'Y'
(...)
#define AXIS_3_NAME 'Z'
(...)
#define AXIS_4_NAME 'Y'
(...)
#define AXIS_5_NAME 'A'
(...)
#define AXIS_6_NAME 'C'
(...)
```  
From your drawing I see that you plan to use externals stepper drivers for your linear axes and ‎DRV8825 drivers or equivalent for rotational axes. This mean you will add a (small and cheap) adapter between the RAMPS board pins on aux3 and your C axis driver.  
  
For auto squaring, you need to connect 2 limits switches on both Y axis. Show [pinout mapping](https://github.com/fra589/grbl-Mega-5X/wiki/grbl-Mega-5X-pinout) for connecting all your switches.
Then use an homing sequence which home both Y axis in the same time. For example:
```(...)
#elif N_AXIS == 6 // 6 axis : homing
  #define HOMING_CYCLE_0 (1<<AXIS_3)  // Home Z axis first to clear workspace.
  #define HOMING_CYCLE_1 ((1<<AXIS_1) // Home X axis
  #define HOMING_CYCLE_2 (1<<AXIS_2)|(1<<AXIS_4) // Home Y axis
  #define HOMING_CYCLE_3 (1<<AXIS_5)  // Home 4th axis (A) (comment if not needed)
  #define HOMING_CYCLE_4 (1<<AXIS_6)  // Home 5th axis (C) (comment if not needed)
#else // Classic 3 axis
(...)
```
Currently, grbl-Mega-5X does not manage offset for auto squaring in software. It's planned in my TODO list for soon, but currently, it will have to be done mechanically by adjusting the position of the switches.
  
There is also no automatic tool change functionality in Grbl, that's up to the Grbl management/GCode sender application to deal with it.
  
@++;  
Gauthier.

**Reply by mlinka** [reply]
Thank you @fra589 .

Sounds great, 
just like you wrote I planned to use cloned Y axis on 6th axis because of DRV8825 sockets that are available on 4th and 5th axis so I wanted to use thoose to get a cleaner look, but adapter you recommended will work if thats the way to do it :)

Thank you, I will also try the homing cycle, it looks great.
I think mechanical settings of auto squarimg switches will work.

Thank you very much for sharing your knowledge.
Best regards.

**Reply by saadbhattii** [reply]
@mlinka Hey, really cool project! I would love to know the updates, did your machine work appropiately and perfectly? Did you come across any other errors? 

Thankyou!
